runtime/criusupport/criusupport.cpp

```diff
@@ -56,110 +56,114 @@ extern "C" {
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_log_file", (UDATA*)&vmCheckpointState->criuSetLogFileFunctionPointerType, "P")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_leave_running", (UDATA*)&vmCheckpointState->criuSetLeaveRunningFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_ext_unix_sk", (UDATA*)&vmCheckpointState->criuSetExtUnixSkFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_file_locks", (UDATA*)&vmCheckpointState->criuSetFileLocksFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_tcp_established", (UDATA*)&vmCheckpointState->criuSetTcpEstablishedFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_auto_dedup", (UDATA*)&vmCheckpointState->criuSetAutoDedupFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_track_mem", (UDATA*)&vmCheckpointState->criuSetTrackMemFunctionPointerType, "Z")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_set_work_dir_fd", (UDATA*)&vmCheckpointState->criuSetWorkDirFdFunctionPointerType, "P")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_init_opts", (UDATA*)&vmCheckpointState->criuInitOptsFunctionPointerType, "V")
+		|| j9sl_lookup_name(vmCheckpointState->libCRIUHandle, (char*)"criu_dump", (UDATA*)&vmCheckpointState->criuDumpFunctionPointerType, "V")
 	) {
-		*currentExceptionClass = criuSystemCheckpointExceptionClass;
-		*systemReturnCode = 1;
-		*nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_CRIU_LOADING_LIBCRIU_FUNCTIONS_FAILED, NULL);
-		returnCode = false;
+		systemReturnCode = 1;
+		nlsMsgFormat = j9nls_lookup_message(J9NLS_DO_NOT_PRINT_MESSAGE_TAG | J9NLS_DO_NOT_APPEND_NEWLINE, J9NLS_JCL_CRIU_LOADING_LIBCRIU_FUNCTIONS_FAILED, NULL);
 		goto done;
 	}
 
 	if ((FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_AGENTLIB_JDWP_EQUALS, NULL) >= 0)
 		|| (FIND_ARG_IN_VMARGS(STARTSWITH_MATCH, MAPOPT_XRUNJDWP, NULL) >= 0)
 	) {
-		vm->checkpointState.flags |= J9VM_CRIU_IS_JDWP_ENABLED;
+		vmCheckpointState->flags |= J9VM_CRIU_IS_JDWP_ENABLED;
 	}
 
 done:
-	return returnCode;
+	if (0 != systemReturnCode) {
+		UDATA msgCharLength = j9str_printf(PORTLIB, NULL, 0, nlsMsgFormat, systemReturnCode);
+		char *exceptionMsg = (char*)j9mem_allocate_memory(msgCharLength, J9MEM_CATEGORY_VM);
+		j9str_printf(PORTLIB, exceptionMsg, msgCharLength, nlsMsgFormat, systemReturnCode);
+		jstring jExceptionMsg = env->NewStringUTF(exceptionMsg);
+		if (JNI_FALSE == env->ExceptionCheck()) {
+			jobject exception = env->NewObject(
+					vmCheckpointState->criuSystemCheckpointExceptionClass, vmCheckpointState->criuSystemCheckpointExceptionInit, jExceptionMsg, (jint)systemReturnCode);
```

Comment:
the checkpoint exceptions are for when we call checkpointJVM, thats the only place where the spec says they can be thrown

Reply:
I think its enough to return false here if it fails

Reply:
Modified `setupJNIFieldIDsAndCRIUAPI()` to return a boolean.
@tajila this is ready for another look.